Orange/widgets/utils/concurrent.py

```diff
@@ -784,3 +782,167 @@ def __call__(self, *args):
+        self.watcher.setFuture(self.future)
+        return self.future
+
+    def cancel(self):
+        assert not self.__interruption_requested
+        self.__interruption_requested = True
+        if self.future is not None:
+            rval = self.future.cancel()
+        else:
+            # not even scheduled yet
+            rval = True
+        return rval
+
+
+class InterruptRequested(BaseException):
+    pass
+
+
+class ConcurrentWidgetMixin:
+    def __init__(self):
+        self.__executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+        self.__task = None  # type: Optional[TaskState]
+
+    @property
+    def task(self):
+        return self.__task
+
+    def _prepare_task(self, state: TaskState) -> Callable[[], Any]:
+        raise NotImplementedError
+
+    def _set_partial_results(self, result: Any) -> None:
+        raise NotImplementedError
+
+    def _set_results(self, results: Any) -> None:
+        # NOTE: All of these have already been set by _set_partial_results,
+        # we double check that they are aliases
+        for key in results.__dict__:
+            value = getattr(results, key)
+            if value is not None:
+                setattr(self, key, value)
+
+    def __set_state_ready(self):
+        self.progressBarFinished()
+        self.setBlocking(False)
+        self.setStatusMessage("")
+
+    def __set_state_busy(self):
+        self.progressBarInit()
+        self.setBlocking(True)
+
+    def __start_task(self, task: Callable[[], Any], state: TaskState):
+        assert self.__task is None
+        state.status_changed.connect(self.setStatusMessage)
+        state.progress_changed.connect(self.progressBarSet)
+        state.partial_result_ready.connect(self._set_partial_results)
+        state.watcher.done.connect(self.on_done)
+        state.start(self.__executor, task)
+        state.setParent(self)
+        self.__task = state
+
+    def __cancel_task(self, wait=True):
+        if self.__task is not None:
+            state, self.__task = self.__task, None
+            state.cancel()
+            state.partial_result_ready.disconnect(self._set_partial_results)
+            state.status_changed.disconnect(self.setStatusMessage)
+            state.progress_changed.disconnect(self.progressBarSet)
+            state.watcher.done.disconnect(self.on_done)
+            if wait:
+                concurrent.futures.wait([state.future])
+                state.deleteLater()
+            else:
+                w = FutureWatcher(state.future, parent=state)
+                w.done.connect(state.deleteLater)
+
+    def start(self):
+        """ Call to start the task. """
+        self.__cancel_task(wait=False)
+
+        if self.data is None:
+            self.__set_state_ready()
+            return
+
+        state = TaskState(self)
+        task = self._prepare_task(state)
+        self.__set_state_busy()
+        self.__start_task(task, state)
+
+    def on_done(self, future: Future):
```

Comment:
I can't think of a single use-case where one would want to override this method. This should probably be private since it deals with futures. The idea of the mixin is to hide all the thread/future specific stuff from us, so even having this available to us to override seems like asking for trouble.

Perhaps we should also rename this to something more explicit e.g. `__on_task_done`.